Azure Programmable Connectivity .NET SDK (#42805)
Azure/azure-sdk-for-net: sdk/programmableconnectivity/Azure.Communication.ProgrammableConnectivity/README.md

# Azure ProgrammableConnectivity client library for .NET

> This SDK allows .NET applications to interact with Azure Programmable Connectivity Gateway, instead of using `curl` requests. For information on APC, see [docs](https://learn.microsoft.com/en-us/azure/programmable-connectivity/).

Reviewer: The "Azure SDK" is the all-up branding for the collection of client and management libraries.   Please use "client library" to refer to this specific library.

Author: Have changed to `This client library allows .NET`...

  [Source code](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/programmableconnectivity/Azure.Communication.ProgrammableConnectivity/src) | [Package (NuGet)](https://www.nuget.org/packages/Azure.Communication.ProgrammableConnectivity) | [API reference documentation](https://azure.github.io/azure-sdk-for-net) | [Product documentation](https://docs.microsoft.com/azure)

> ## Getting started

Reviewer: ```suggestion
## Getting started

```

### Install the package

Install the client library for .NET with [NuGet](https://www.nuget.org/):

```dotnetcli
dotnet add package Azure.Communication.ProgrammableConnectivity --prerelease
```

### Prerequisites

Before you can use the SDK successfully, you'll need to

> * Follow the [guide](https://learn.microsoft.com/en-us/azure/programmable-connectivity/azure-programmable-connectivity-create-gateway) to create a gateway, or have one already.

Reviewer: ```suggestion
* Follow the [guide](https://learn.microsoft.com/azure/programmable-connectivity/azure-programmable-connectivity-create-gateway) to create a gateway, or have one already.
```
nit:  You'll need to remove the language slugs from the MS Learn links.

Author: thank you for pointing out!

> * Note down your endpoint and `apc-gateway-id`.

Reviewer 2: **Suggestion:** As much as it feels obvious, it never hurts to be super explicit about where to get this (in this case you can just say following the guide linked in the previous bulletpoint)

### Authenticate the client

> The SDK uses `DefaultAzureCredential()` to authenticate with APC. See [docs](https://learn.microsoft.com/en-us/dotnet/api/azure.identity.environmentcredential?view=azure-dotnet) for information on how to allow this to access your credentials.

Reviewer 2: **Question:** Does the SDK actually specify what type of authentication has to be used? Is it not just any Azure auth?

Author: That's a good question. I believe it works with any kind. In the python sdk it definitely works with any kind.

Author: have updated the docs.

> The SDK uses `DefaultAzureCredential()` to authenticate with APC. See [docs](https://learn.microsoft.com/en-us/dotnet/api/azure.identity.environmentcredential?view=azure-dotnet) for information on how to allow this to access your credentials.

Reviewer: This isn't quite correct; `DefaultAzureCredential` is just one of the available Azure.Identity credentials.  The client library accepts all of them, not just DAC.

Author: Thank you, I missed this

## Key concepts

> For each call that you make to APC with the SDK, you will follow the same pattern:

Reviewer 2: **Suggestion:** Write this as a real code snippet with comments. For example you haven't defined `baseClient` in the bulletpoints below

Author: I'll make them a bit clearer, and mention that the reader can have a look at the `Examples` folder

* Create a client `ProgrammableConnectivityClient`
* Access the sub-client for your use case (sim-swap/location/number-verification/device-network) by calling say `baseClient.GetSimSwapClient()`
* Create the content for your request by using the objects given by the SDK, for example `SimSwapVerificationContent`
* Call the client with the content you've created
* Access the result

## Examples

You can familiarize yourself with different APIs using [Samples](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/programmableconnectivity/Azure.Communication.ProgrammableConnectivity/samples).

## Troubleshooting

> If your call doesn't work, we recommend logging out the exception messages, and progressing from there. You can see an example of how to handle errors [here](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/programmableconnectivity/Azure.Communication.ProgrammableConnectivity/samples/README.md)

Reviewer: ```suggestion
If your call doesn't work, we recommend logging the exception messages, and progressing from there. You can see an example of how to handle errors [here](https://github.com/Azure/azure-sdk-for-net/tree/main/sdk/programmableconnectivity/Azure.Communication.ProgrammableConnectivity/samples/README.md)
```

## Contributing

> APC is currently not accepting/excpecting contributions for this codedbase.

Reviewer 2: **Typo:**
```suggestion
APC is currently not accepting/expecting contributions for this codebase.
```

# Azure ProgrammableConnectivity client library for .NET

> This section should give out brief introduction of the client library.

Author: changed the README.md
